devito/grid.py

```diff
@@ -0,0 +1,50 @@
+from devito.tools import as_tuple
+from devito.dimension import x, y, z
+
+import numpy as np
+
+__all__ = ['Grid']
+
+
+class Grid(object):
+
+    """
+    A cartesian grid that encapsulates a physical domain over which
+    to discretize :class:`Function`s.
+
+    :param shape: Shape of the domain region in grid points.
+    :param extent: Physical extent of the domain in m; defaults to a
+                   unit box of extent 1m in all dimensions.
+    :param origin: Physical coordinate of the origin of the domain;
+                   defaults to 0. in all dimensions.
+    :param dtype: Default data type to be inherited by all Functions
+                  created from this :class:`Grid`.
+    """
+
+    def __init__(self, shape, extent=None, origin=None, dimensions=None,
+                 dtype=np.float32):
+        self.shape = as_tuple(shape)
+        self.extent = as_tuple(extent or tuple(1. for _ in shape))
+        self.origin = as_tuple(origin or tuple(0. for _ in shape))
+        self.dtype = dtype
+
+        # TODO: Raise proper exceptions and logging
+        assert(self.dim == len(self.origin) == len(self.extent) == len(self.spacing))
+
+        # TODO: Create Dimensions locally instead of using global ones
+        self.dimensions = (dimensions or (x, y, z))[:self.dim]
+
+    @property
+    def dim(self):
+        """Problem dimension, or number of spatial dimensions."""
+        return len(self.shape)
+
+    @property
+    def spacing(self):
+        """Spacing between grid points in m."""
+        return as_tuple(np.array(self.extent) / np.array(self.shape))
```

Comment:
`(np.array(self.extent) - np.array(self.origin)) / (np.array(self.shape) -1)`
if kept with origin/extend/shape

Reply:
As mentioned above, I think we have different interpretations of `extent`. I see it purely as the "width" of the domain, not the top-right-back coordinate.

Reply:
The I would call it `width`, extent sounds like "how far" and needs to know the origin to know the actual size, the "-1" is till required

Reply:
Yeah no, `extent` includes, for example in 3D, the height of the grid, and generally speaking `width != height`. The point about extent is that it is, as far as I interpret the word usually, a size relative to an origin - and since our domain has an "origin", the extent (relative size) of the domain in combination with the absolute coordinates of the origin defines the entire physical space that our domain covers. 

In fact, in your interpretation "extent" seems to be an absolute coordinate, and I'm not entirely sure how you justify that?

Reply:
This are more practical considerations. In practice two things happen:

1- People choose a `space_order` according to some empirical/theoretical results on dispersion and from it, choose the optimal spacing, so that is the parameter they will know/want to input

2- The actual computational domain is defined by physical constraints such as in seismic case, what is the maximum/minimum src/rec position, so the "extent" the way you define it is not the known, origin/end is what is known.

Which is why the usual origin/spacing + either end/npoints is the convention.

Reply:
Ah ok, now this is starting to make sense. The problem though is that this is a somewhat specific convention in seismic, I believe. As far as I'm aware, more generic physics modelling (eg. CFD) starts from some description of the physical domain, which might have geometric features like inlets/outlets on a channel, and discretizes this to arrive at the grid spacing. 

Since we want to keep Devito a generic FD engine, I would therefore argue we keep this convention on the `Grid`, while the `Model`, which is clearly specific to seismic cases, implements the convention you described. After all it is quite trivial to switch between the two.

Reply:
>As far as I'm aware, more generic physics modelling (eg. CFD) starts from some description of the physical domain, which might have geometric features like inlets/outlets on a channel, and discretizes this to arrive at the grid spacing.

The grid spacing would still be constrained by the discretization order, but fine with me I guess

Reply:
still needs `(np.array(self.shape) -1)` to get the correct spacing.

Reply:
True, just pushed.